uclibc-ng: improve detection of FPU, endianess and ABI

aleclearmind opened the pull request:
This branch improves the support for building toolchains using `uclibc-ng`.

Currently, uclibc-ng seems to have harcoded support for certain configurations, in particular for ARM.

This branch makes the package more generic and automatically configures uclibc-ng in terms of endianess, availability of an FPU and ABI of choice.

The branch also correctly initializes the configuration by specifying the target architecture upon `make defconfig`.

The last commit reports a simple script to test certain toolchains for MIPS and ARM.

Note: this is my first PR, any feedback is highly appreciated.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

aleclearmind committed:
uclibc-ng: minor changes

aleclearmind committed:
uclibc-ng: do not disable hardware floating point support for ARM

The original author had probably in mind some specific device without
hardware floating point support.

aleclearmind committed:
uclibc-ng: detect CONFIG_ARM_EABI from platform

This commit introduces a `isEabi` predicate for platforms and employs it
to detect whether to configure uclibc with CONFIG_ARM_EABI or not.

aleclearmind committed:
uclibc-ng: handle endianess explicitly

Employ `isLittleEndian`/`isBigEndian` to configure uclibc endianess.

aleclearmind committed:
uclibc-ng: specify architecture in `make defconfig`

This ensures the correct architecture is picked up in the `.config` file.
For instance, specifying `ARCH=arm` will correctly set TARGET_ARCH=arm
and TARGET_arm=y.

This commit has been tested to successfully produce toolchains for:

* mipsel-unknown-linux-uclibc (ben-nanonote)
* mips-unknown-linux-uclibc
* armv7a-unknown-linux-uclibceabihf
* armv7a-unknown-linux-uclibceabi

Specifically, the following script has been used to test the toolchains:

    #!/usr/bin/env bash

    set -euo pipefail

    function test_toolchain() {
        TRIPLE="$1"
        shift
        EXPRESSION="$1"
        shift
        QEMU_ARCH="$1"
        shift

        echo "Testing $TRIPLE" > /dev/stderr

        BUILT_PACKAGE=$(nix-build -E "$EXPRESSION")
        nix-shell -p "$BUILT_PACKAGE" -p qemu --run '
          set -euo pipefail

          echo "#include <stdio.h>" > program.c
          echo "int main() { puts(\"Hello!\"); }" >> program.c
          '"$TRIPLE"'-gcc program.c -o program -static

          for TO_GREP in '"$(printf "'%s' " "$@")"'; do
              readelf -h program | grep -F "$TO_GREP"
          done

          qemu-'"$QEMU_ARCH"' ./program
        '
    }

    function test_triple() {
        TRIPLE="$1"
        shift
        QEMU_ARCH="$1"
        shift

        test_toolchain "$TRIPLE" '(import ./. { crossSystem = { config = "'"$TRIPLE"'"; }; }).stdenv.cc' "$QEMU_ARCH" "$@"
    }

    function test_cross() {
        TRIPLE="$1"
        shift
        CROSS="$1"
        shift
        QEMU_ARCH="$1"
        shift

        test_toolchain "$TRIPLE" '(import ./. {}).pkgsCross.'"$CROSS"'.stdenv.cc' "$QEMU_ARCH" "$@"
    }

    test_cross "mipsel-unknown-linux-uclibc" "ben-nanonote" "mipsel" "little endian"
    test_triple "mips-unknown-linux-uclibc" "mips" "big endian"
    test_triple "armv7a-unknown-linux-uclibceabihf" "arm" "little endian" "EABI" "hard-float ABI"
    test_triple "armv7a-unknown-linux-uclibceabi" "arm" "little endian" "EABI" "soft-float ABI"

aleclearmind committed:
uclibc-ng: use V=1 for verbose builds

VERBOSE=1 is no (longer?) used by uclibc.

aleclearmind committed:
all-packages: drop reference to obsolete uclibc package in favor of uclibc-ng

aleclearmind commented:
@rasendubi any chance you can take a look at this?

aleclearmind commented on pkgs/by-name/uc/uclibc-ng/package.nix:
```diff
@@ -51,8 +51,10 @@ let
+  + lib.optionalString (lib.hasInfix "uclibceabi" stdenv.hostPlatform.config) ''
     CONFIG_ARM_EABI y
+  ''
+  + lib.optionalString (stdenv.hostPlatform.isAarch32) ''
```
To be honest I have no idea why `isCross` was there. The oldest reference to check if we're cross compiling dates back to d65fe65616f28b60102fd9d05e593676e5a75555.

The whole handling of `uclibc` for ARM seemed to be targeting a very specific device the original developer had in mind.

Shall I add a comment on the commit message on the line "This commit also remove isCross which was there for unclear reasons."?

alyssais replied:
Separate commit is even better in that sort of situation because it means your other changes are less likely to be reverted if the confusing one turns out to cause a problem. (Although with uclibc I doubt there are enough users for that to happen regardless, so this is more of a general good practice.)

aleclearmind replied:
Moved to a separate commit.

aleclearmind committed:
uclibc-ng: do not special case ARM cross-compilation

It is unclear why this was being done.
The earliest source of detecting cross-compilation for ARM dates back to
d65fe65.

aleclearmind committed:
uclibc-ng: add aleclearmind as maintainer

aleclearmind committed:
maintainers: add aleclearmind

alyssais commented on pkgs/by-name/uc/uclibc-ng/package.nix:
```diff
@@ -51,8 +51,10 @@ let
+  + lib.optionalString stdenv.hostPlatform.isEabi ''
     CONFIG_ARM_EABI y
+  ''
+  + lib.optionalString (stdenv.hostPlatform.isAarch32) ''
```
No need for the parens here.

alyssais reviewed (commented):
Looking good!  Merge conflict needs fixing, and it would be nice to see more reasoning for each change in the commit messages.

aleclearmind commented:
> Merge conflict needs fixing,

Done.

> it would be nice to see more reasoning for each change in the commit messages

I added some more reasoning, but certain things would mostly delve into guessing what the original author intended to do, which is a bit hard.

Ericson2314 commented:
This PR looks very good! Maybe we can even delete the non-NG uclibc soon too.

aleclearmind commented on lib/systems/inspect.nix:
```diff
@@ -437,9 +437,25 @@ rec {
+        musleabihf
+        uclibceabi
+        uclibceabihf
+      ];
```
Done. Note that I followed the current practice (see `isGnu`, `isMusl`).

Ericson2314 reviewed (approved):
Let's wait for @alyssais, but I think this looks good to me!

alyssais commented:
Thank you!
